src/daemon/Display.cpp

```diff
@@ -370,52 +396,120 @@ namespace SDDM {
             return;
+        }
+
+        // failed login only when result not PAM_SUCCESS,
+        // i.e. error is of type ERROR_AUTHENTICATION
+        if (error == AuthEnums::ERROR_AUTHENTICATION)
+            emit loginFailed(m_socket, message, result);
+        else if(error == AuthEnums::ERROR_PAM_CONV)
+                emit pamConvMsg(m_socket, message, result);
 
-        if (error == Auth::ERROR_AUTHENTICATION)
-            emit loginFailed(m_socket);
+        // ignore internal errors, no emit
     }
 
-    void Display::slotHelperFinished(Auth::HelperExitStatus status) {
+    void Display::slotHelperFinished(AuthEnums::HelperExitStatus status) {
         // Don't restart greeter and display server unless sddm-helper exited
         // with an internal error or the user session finished successfully,
         // we want to avoid greeter from restarting when an authentication
         // error happens (in this case we want to show the message from the
         // greeter
-        if (status != Auth::HELPER_AUTH_ERROR)
+        if (status != AuthEnums::HELPER_AUTH_ERROR)
             stop();
     }
 
+    /* got new request (with prompts) from pam conv(), e.g. for expired pwd
+     * which requires response from greeter UI (new password) */
     void Display::slotRequestChanged() {
-        if (m_auth->request()->prompts().length() == 1) {
-            m_auth->request()->prompts()[0]->setResponse(qPrintable(m_passPhrase));
-            m_auth->request()->done();
-        } else if (m_auth->request()->prompts().length() == 2) {
-            m_auth->request()->prompts()[0]->setResponse(qPrintable(m_auth->user()));
-            m_auth->request()->prompts()[1]->setResponse(qPrintable(m_passPhrase));
-            m_auth->request()->done();
+
+        int n_prompts = m_auth->request()->prompts().count();
+
+        // ignore empty requests
+        if(n_prompts<=0) return;
+
+        // see what we got from pam conv() and will be send to greeter
+        qDebug() << "Display: requestChanged with " << n_prompts << " prompts from Auth";
+
+        // handle password change case (gets response from greeter),
+        // will finish with request->done() later in setPamResponse()
+        if (m_auth->request()->findPrompt(AuthPrompt::CHANGE_PASSWORD))
+        {
+            if(m_socket)
+                // send password change request to greeter (via SocketServer)
+                emit pamRequest(m_socket, m_auth->request());
+            return;
+        }
+
+        // handle requests with user name and password prompts, respond to login password request
+        if ((m_auth->request()->findPrompt(AuthPrompt::LOGIN_USER) ||
+             m_auth->request()->findPrompt(AuthPrompt::LOGIN_PASSWORD)) &&
+             m_auth->request()->setLoginResponse(m_auth->user(), m_passPhrase) == true)
+        {
+             m_auth->request()->done();
+             return;
         }
+
+        qWarning() << "Display: Unable to handle Auth request!";
```

Comment:
Cancel auth requests which cannot be handled. Otherwise clicking the Login button with an empty username field would prevent subsequent login attempts:
```
m_auth->request()->canceled();
````

Reply:
Sure, makes sense. I can add it.

Reply:
Done.